File: apps/frontend/src/lib/types/collection.ts
import type { TrackResponse } from './track';

export interface CollectionResponse {
	id: number;
	name: string;
	description: string | null;
	user_id: number;
	is_public: boolean;
	created_at: string;
	updated_at: string;
}

export interface CollectionTrackResponse {
	id: number;
	collection_id: number;
	track_id: number;
	position: number;
	added_at: string;
	track: TrackResponse;
}

export interface CollectionDetailResponse extends CollectionResponse {
	tracks: CollectionTrackResponse[];
}

export interface CollectionListResponse {
	items: CollectionResponse[];
	total: number;
	page: number;
	page_size: number;
}

export interface CollectionCreate {
	name: string;
	description?: string | null;
	is_public?: boolean;
}

export interface CollectionUpdate {
	name?: string | null;
	description?: string | null;
	is_public?: boolean | null;
}

export interface AddTrackToCollection {
	track_id: number;
	position?: number | null;
}

export interface UpdateTrackPosition {
	position: number;
}

export interface ReorderTracksRequest {
	track_ids: number[];
}

export interface CollectionListParams {
	user_id?: number;
	page?: number;
	page_size?: number;
}

File: apps/backend/src/backend/schemas/collection.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from backend.schemas.track import TrackResponse


class CollectionCreate(BaseModel):
    name: str = Field(..., max_length=255)
    description: str | None = None
    is_public: bool = False


class CollectionUpdate(BaseModel):
    name: str | None = Field(None, max_length=255)
    description: str | None = None
    is_public: bool | None = None


class CollectionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None
    user_id: int
    is_public: bool
    created_at: datetime
    updated_at: datetime


class CollectionTrackResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    collection_id: int
    track_id: int
    position: int
    added_at: datetime
    track: TrackResponse


class CollectionDetailResponse(CollectionResponse):
    tracks: list[CollectionTrackResponse] = []


class CollectionListResponse(BaseModel):
    items: list[CollectionResponse]
    total: int
    page: int
    page_size: int


class AddTrackToCollection(BaseModel):
    track_id: int
    position: int | None = None


class UpdateTrackPosition(BaseModel):
    position: int = Field(..., ge=0)


class ReorderTracksRequest(BaseModel):
    track_ids: list[int]
